### ISSUE TITLE
dependecy on BayesLogit for MacOS

### ISSUE THREAD

**Issue by FabianRoger** [hmsc-r/HMSC#11]
This is not an Issue *per se* but just to let you know that the installation of BayseLogit can be very cubersome on OSX. This has to do with the need of compiling the package manually and problems with rcpp on mac afaik. 

I made it work after ~2h following the advice below: 

[library not found for -lgfortran](https://thecoatlessprofessor.com/programming/openmp-in-r-on-os-x/)
[macOS 'wchar.h' File Not Found](https://stackoverflow.com/a/55465527/1183277)

I saw that @jarioksa tried to fix the problems with BayseLogit so it could be resubmitted to CRAN but the author seems to be unresponsive. It's a pity because users might be putt off because of the trouble that ensues. 

___________
    # excerpt from session_info(): 

    version  R version 3.5.3 (2019-03-11)
    os       macOS Mojave 10.14          
    system   x86_64, darwin15.6.0 1
__________

ps: now that I managed to install it anyway, I am really looking forward to start using it

**Comment by gtikhonov** [comment]
There has already been this discussion on issues with OSX, but unfortunately we lost that issue tread when ported the repo from my profile to here. Here is the info:

1. By aksulehikoinen:

> When trying to install this line
install_url('https://cran.r-project.org/src/contrib/Archive/BayesLogit/BayesLogit_0.6.tar.gz')
I got "Install R Tools"-note:
"Building R package from source requires installation of additional build tools."
I have mac, R version 3.5.0 and R-Studio 1.1.453
After installing the Command Line Developer Tools from here
https://developer.apple.com/download/more/
I retried to install the line
install_url('https://cran.r-project.org/src/contrib/Archive/BayesLogit/BayesLogit_0.6.tar.gz')
The installation started but got errors:
ld: warning: directory not found for option '-L/usr/local/gfortran/lib/gcc/x86_64-apple-darwin15/6.1.0'
ld: warning: directory not found for option '-L/usr/local/gfortran/lib'
ld: library not found for -lgfortran
clang: error: linker command failed with exit code 1 (use -v to see invocation)
make: *** [BayesLogit.so] Error 1
ERROR: compilation failed for package ‘BayesLogit’
    removing ‘/Library/Frameworks/R.framework/Versions/3.5/Resources/library/BayesLogit’
    Installation failed: Command failed (1)

2. By aksulehikoinen:

> I downloaded the latest versions of clang and gfortran from
https://cran.r-project.org => Development Tools and Libraries
This helped and I managed to install rest of the packages including Hmsc

3. By taoyun  

> For Mac users who need to update clang and gfortran using .pkg, the aforementioned Development Tools site is here: https://cran.r-project.org/bin/macosx/tools/. This solution worked like a charm.

I hope these can facilitate the installation process for other MacOS users.

**Comment by jarioksa** [comment]
I did my testing in macOS and never noticed there was a problem with building the package. You only need Fortran and to get it, you need to follow the instructions in the R for Mac OS X (still old name) page of CRAN. You also need macOS command line tools and again you need to follow the CRAN instructions. I had done all this for other purposes, and therefore I did not notice any problems in building the package.

**Comment by jarioksa** [comment]
If you use official macOS command line tools, you may need to set a path to library headers. This is documented in R, but the documentation is well hidden: See section C.3 in R Installation and Administration manual https://cran.r-project.org/doc/manuals/r-release/R-admin.html#macOS

**Comment by diogoprov** [comment]
Yes, I checked all the suggestions in the two links by @FabianRoger and still couldn't get BayesLogit to compile. I have gfortran, XCode, and clang installed. I checked homebrew also and puf, nothing seemed to work. 

```
* installing to library ‘/Users/diogoborgesprovete/Library/R/3.6/library’
* installing *source* package ‘BayesLogit’ ...
** package ‘BayesLogit’ successfully unpacked and MD5 sums checked
** using staged installation
** libs
clang++ -I"/Library/Frameworks/R.framework/Resources/include" -DNDEBUG -DUSE_R -DNDEBUG -DDISABLE_SINGLE -DNTHROW   -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX.sdk -I/usr/local/include  -fPIC  -Wall -g -O2  -c Matrix.cpp -o Matrix.o
clang: warning: no such sysroot directory: '/Library/Developer/CommandLineTools/SDKs/MacOSX.sdk' [-Wmissing-sysroot]
In file included from Matrix.cpp:1:
In file included from ./Matrix.h:37:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/vector:265:
In file included from /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/iosfwd:90:
/Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/include/c++/v1/wchar.h:119:15: fatal error: 
      'wchar.h' file not found
#include_next <wchar.h>
              ^~~~~~~~~
1 error generated.
make: *** [Matrix.o] Error 1
ERROR: compilation failed for package ‘BayesLogit’
* removing ‘/Users/diogoborgesprovete/Library/R/3.6/library/BayesLogit’
```

while trying in Terminal with R CMD install BayesLogit, this error happens: 

```
Error: package or namespace load failed for ‘BayesLogit’ in dyn.load(file, DLLpath = DLLpath, ...):
 impossible to load shared object
 '/Users/diogoborgesprovete/Library/R/3.6/library/00LOCK-BayesLogit/00new/BayesLogit/libs/BayesLogit.so':
  dlopen(/Users/diogoborgesprovete/Library/R/3.6/library/00LOCK-BayesLogit/00new/BayesLogit/libs/BayesLogit.so, 6): Symbol not found: ___addtf3
  Referenced from: /usr/local/lib/libquadmath.0.dylib
  Expected in: /usr/local/lib/libgcc_s.1.dylib
 in /usr/local/lib/libquadmath.0.dylib
Erro: loading failed
Execução interrompida
ERROR: loading failed
* removing ‘/Users/diogoborgesprovete/Library/R/3.6/library/BayesLogit’
```

Any ideas?

**Comment by jarioksa** [comment]
The first error message comes from a failure to find header file `wchar.h`. My previous message had a link to R-admin manual that tells you how to fix this. Briefly, macOS does not install header files for command line tools in default settings, but you must install them separately using something like
```
sudo installer -pkg /Library/Developer/CommandLineTools/Packages/macOS_SDK_headers_for_macOS_10.14.pkg -target /
```
(possibly adapting this command if you are not running macOS 10.14.x). If you update your commandline tools, you must run again this command, because re-installing the toolset removes the header files.

**Comment by diogoprov** [comment]
Yes, it didn't work either. I had tried that before, following your comment. The second error still happens now. I'm using macOS 10.14.5 and R 3.6.1.

**Comment by diogoprov** [comment]
```
Error: package or namespace load failed for ‘BayesLogit’ in dyn.load(file, DLLpath = DLLpath, ...):
 unable to load shared object '/Users/diogoborgesprovete/Library/R/3.6/library/00LOCK-BayesLogit/00new/BayesLogit/libs/BayesLogit.so':
  dlopen(/Users/diogoborgesprovete/Library/R/3.6/library/00LOCK-BayesLogit/00new/BayesLogit/libs/BayesLogit.so, 6): Symbol not found: ___addtf3
  Referenced from: /usr/local/lib/libquadmath.0.dylib
  Expected in: /usr/local/lib/libgcc_s.1.dylib
 in /usr/local/lib/libquadmath.0.dylib
Error: loading failed
Execution halted
ERROR: loading failed
* removing ‘/Users/diogoborgesprovete/Library/R/3.6/library/BayesLogit’
Error: Failed to install 'unknown package' from URL:
  (converted from warning) installation of package ‘/var/folders/kd/_k75gv855qxbyrqq0rtqk1w40000gr/T//RtmpDedMtN/file73a643038576/BayesLogit_0.6.tar.gz’ had non-zero exit status
```

**Comment by jarioksa** [comment]
`addtf3` is a standard **gcc** (gnu C compiler) and **clang** function for floating point emulation (that you don't need in Mac, as it has a floating point processor). What actually is your toolset? That is, do your version of build tools *exactly* accord with the instructions in the CRAN front page for Mac? I use macOS 10.14.6 and R 3.6.1, and also regularly build R-devel from the source **and** also use the **BayesLogit**  built from the source and I never experienced a problem. On the other hand, I would not touch homebrew tools with a six-feet stick and it does not mess my system.

**Comment by dougwyu** [comment]
My student is having trouble installing BayesLogit on macOS 10.14.6.  I thought her system was the same as mine (I have no problem), but apparently it's not, and I don't know how to interpret these error messages. We did run the macos_SDK_headers installation command, and it installed correctly. 

```
library(devtools)
install_url('https://cran.r-project.org/src/contrib/Archive/BayesLogit/BayesLogit_0.6.tar.gz')
```
```
Warning in install.packages :
  unable to access index for repository jarioksa.github.io/drat/bin/macosx/el-capitan/contrib/3.5:
  cannot open URL 'jarioksa.github.io/drat/bin/macosx/el-capitan/contrib/3.5/PACKAGES'
Package which is only available in source form, and may need compilation of
  C/C++/Fortran: ‘BayesLogit’
Do you want to attempt to install these from sources? (Yes/no/cancel) Y
installing the source package ‘BayesLogit’

trying URL 'jarioksa.github.io/drat/src/contrib/BayesLogit_0.6-1.tar.gz'
Content type 'application/gzip' length 283153 bytes (276 KB)
downloaded 276 KB

* installing *source* package ‘BayesLogit’ ...
** libs
clang++  -I"/Library/Frameworks/R.framework/Resources/include" -DNDEBUG -DUSE_R -DNDEBUG -DDISABLE_SINGLE -DNTHROW   -I/usr/local/include   -fPIC  -Wall -g -O2  -c Matrix.cpp -o Matrix.o
In file included from Matrix.cpp:1:
In file included from ./Matrix.h:37:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/vector:269:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/iosfwd:90:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/wchar.h:119:
/usr/include/wchar.h:98:15: error: unknown type name 'FILE'
wint_t  fgetwc(FILE *);
               ^
/usr/include/wchar.h:99:44: error: unknown type name 'FILE'
wchar_t *fgetws(wchar_t * __restrict, int, FILE * __restrict);
                                           ^
/usr/include/wchar.h:100:24: error: unknown type name 'FILE'
wint_t  fputwc(wchar_t, FILE *);
                        ^
/usr/include/wchar.h:101:40: error: unknown type name 'FILE'
int     fputws(const wchar_t * __restrict, FILE * __restrict);
                                           ^
/usr/include/wchar.h:102:11: error: unknown type name 'FILE'
int     fwide(FILE *, int);
              ^
/usr/include/wchar.h:103:14: error: unknown type name 'FILE'
int     fwprintf(FILE * __restrict, const wchar_t * __restrict, ...);
                 ^
/usr/include/wchar.h:104:13: error: unknown type name 'FILE'
int     fwscanf(FILE * __restrict, const wchar_t * __restrict, ...);
                ^
/usr/include/wchar.h:105:14: error: unknown type name 'FILE'
wint_t  getwc(FILE *);
              ^
/usr/include/wchar.h:113:23: error: unknown type name 'FILE'
wint_t  putwc(wchar_t, FILE *);
                       ^
/usr/include/wchar.h:117:24: error: unknown type name 'FILE'
wint_t  ungetwc(wint_t, FILE *);
                        ^
/usr/include/wchar.h:118:15: error: unknown type name 'FILE'
int     vfwprintf(FILE * __restrict, const wchar_t * __restrict,
                  ^
/usr/include/wchar.h:170:14: error: unknown type name 'FILE'
int     vfwscanf(FILE * __restrict, const wchar_t * __restrict,
                 ^
/usr/include/wchar.h:205:1: error: unknown type name 'FILE'
FILE *open_wmemstream(wchar_t ** __bufp, size_t * __sizep) __API_AVAILABLE(macos(10.13), ios(11.0), tvos(11.0), watchos(4.0));
^
/usr/include/wchar.h:215:18: error: unknown type name 'FILE'
wchar_t *fgetwln(FILE * __restrict, size_t *) __OSX_AVAILABLE_STARTING(__MAC_10_7, __IPHONE_4_3);
                 ^
In file included from Matrix.cpp:1:
In file included from ./Matrix.h:38:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/iostream:38:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/ios:216:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/__locale:15:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/string:500:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/string_view:176:
In file included from /Library/Developer/CommandLineTools/usr/include/c++/v1/__string:57:
/Library/Developer/CommandLineTools/usr/include/c++/v1/cstdio:108:9: error: no member named 'FILE' in the global namespace
using ::FILE;
      ~~^
/Library/Developer/CommandLineTools/usr/include/c++/v1/cstdio:109:9: error: no member named 'fpos_t' in the global namespace
using ::fpos_t;
      ~~^
/Library/Developer/CommandLineTools/usr/include/c++/v1/cstdio:112:9: error: no member named 'fclose' in the global namespace
using ::fclose;
      ~~^
/Library/Developer/CommandLineTools/usr/include/c++/v1/cstdio:113:9: error: no member named 'fflush' in the global namespace
using ::fflush;
      ~~^
/Library/Developer/CommandLineTools/usr/include/c++/v1/cstdio:114:9: error: no member named 'setbuf' in the global namespace
using ::setbuf;
      ~~^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
make: *** [Matrix.o] Error 1
ERROR: compilation failed for package ‘BayesLogit’
* removing ‘/Library/Frameworks/R.framework/Versions/3.5/Resources/library/BayesLogit’
Warning in install.packages :
  installation of package ‘BayesLogit’ had non-zero exit status
```

**Comment by jarioksa** [comment]
@dougwyu **BayesLogit** is now in **CRAN**. You can install it in usual way. That is, with `install.packages("BayesLogit")` on command line or using the pulldown menus in R GUI. This is also now told in `README.md` of **Hmsc**. 

The problem was that your student had **R** 3.5.x and we had not made binary available for that older version of **R** simply because we did not have the old version. However, **CRAN** has **macOS** binary also for **R** 3.5.x (but different version than for the current **R**, but both work similarly). So use **CRAN** instead of the old guidance.

**Comment by diogoprov** [comment]
Just to confirm that I was able to get it installed in R 3.6.1 using macOS 10.14.6 by installing BayesLogit from CRAN.

Thank you.

P.S.: the v 3.0-1 comes without vignettes.

**Comment by jarioksa** [comment]
**Hmsc** is now in CRAN (not all binaries are yet available when writing this, but they will be within some days). **BayesLogit** is also in CRAN. This means that you can install **Hmsc** as any other **R** package. The only complication is that **BayesLogit** is not available for **R** older than 3.6.0 in Windows (macOS is OK), and therefore **Hmsc** cannot be installed unless you first install **BayesLogit** by some other means (like through `repo` https://jarioksa.github.io/drat).
